miniapps/solvers/schrodinger_flow.hpp

```diff
@@ -0,0 +1,947 @@
+   void GaugeTransform()
+   {
+      const auto q_r = Reshape(q.Read(), ndofs);
+      auto psi1_r = Reshape(psi1.real().ReadWrite(), ndofs);
+      auto psi1_i = Reshape(psi1.imag().ReadWrite(), ndofs);
+      auto psi2_r = Reshape(psi2.real().ReadWrite(), ndofs);
+      auto psi2_i = Reshape(psi2.imag().ReadWrite(), ndofs);
+      mfem::forall(ndofs, [=] MFEM_HOST_DEVICE(int n)
+      {
+         const complex_t minus_i(0, -1.0);
+         const complex_t eiq = exp(minus_i * q_r(n));
+         complex_t psi1(psi1_r(n), psi1_i(n)), psi2(psi2_r(n), psi2_i(n));
+         psi1 *= eiq, psi2 *= eiq;
+         psi1_r(n) = psi1.real(), psi1_i(n) = psi1.imag();
+         psi2_r(n) = psi2.real(), psi2_i(n) = psi2.imag();
+      });
+   }
+
+   /// @brief Add a circular vortex to the wavefunctions psi1 and psi2.
+   void AddCircularVortex(const real3_t center, const real3_t normal,
+                          const real_t radius, const real_t swirling)
+   {
+      MFEM_VERIFY(swirling > 0.0, "Swirling strength must be positive");
+      const auto DIM = dim;
+      const real3_t o = { center[0], center[1], center[2] };
+      const real_t norm2 = std::sqrt(normal[0] * normal[0] +
+                                     normal[1] * normal[1] +
+                                     normal[2] * normal[2]);
+      const real_t n0 = normal[0] / norm2, n1 = normal[1] / norm2,
+                   n2 = normal[2] / norm2;
+      const auto &X = Reshape(nodes.Read(), dim, ndofs);
+      auto psi1_r = Reshape(psi1.real().ReadWrite(), ndofs);
+      auto psi1_i = Reshape(psi1.imag().ReadWrite(), ndofs);
+      mfem::forall(ndofs, [=] MFEM_HOST_DEVICE(int n)
+      {
+         const real_t px = X(0, n), py = X(1, n),
+                      pz = DIM == 3 ? X(2, n) : 0.0;
+         const real_t rx = px - o[0], ry = py - o[1], rz = pz - o[2];
+         const real_t z = rx * n0 + ry * n1 + rz * n2;
+         const bool inRange = (rx * rx + ry * ry + rz * rz - z * z) < radius * radius;
+         const bool inLayerP = inRange && (z > 0.0  && z <= (+swirling / 2.0));
+         const bool inLayerM = inRange && (z <= 0.0 && z >= (-swirling / 2.0));
+         real_t alpha = 0.0;
+         if (inLayerP) { alpha = -M_PI * (2.0 * z / swirling - 1.0); }
+         if (inLayerM) { alpha = -M_PI * (2.0 * z / swirling + 1.0); }
+         complex_t psi1(psi1_r(n), psi1_i(n));
+         const complex_t alpha_i(0, alpha);
+         psi1 *= exp(alpha_i);
+         psi1_r(n) = psi1.real(), psi1_i(n) = psi1.imag();
+      });
+   }
+};
+
+/// @brief Crank-Nicolson time solver for the Schrodinger equation.
+template<typename TFiniteElementSpace,
+         typename TSesquilinearForm,
+         typename TComplexGridFunction>
+struct CrankNicolsonBaseSolver
+{
+   ConstantCoefficient one, dthq, mdthq;
+   TSesquilinearForm C_form, R_form;
+   Array<int> no_bc;
+   OperatorHandle C_op, R_op;
+   TComplexGridFunction z;
+   GMRESSolver Cm1;
+
+   // ∂ₜ𝝭 = ½ℏ𝑖∆𝝭
+   // 𝝭ⁿ⁺¹ - 𝝭ⁿ = ¼𝛅t𝑖ℏ(∆𝝭ⁿ⁺¹ + ∆𝝭)
```

Comment:
I do not have fancy unicode knowledge, so I apologize this is not a GitHub suggestion... but I believe the last $\Delta \Psi$ should be $\Delta \Psi^n$